### PR TITLE
Make numpy an optional dependency in `utilities\seed.py`

### PR DESCRIPTION
**Pull request by 01AbhiSingh**
## What does this PR do?

Part of #16649

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Numpy no longer a required dependency in ```utilites\seed.py```.
Made the following changes in code : 

1. Added RequirementCache:
```from lightning_utilities.core.imports import RequirementCache
_NUMPY_AVAILABLE = RequirementCache("numpy")
```

2. Fallback for max_seed_value and min_seed_value:
```
if _NUMPY_AVAILABLE:
    import numpy as np
    max_seed_value = np.iinfo(np.uint32).max
    min_seed_value = np.iinfo(np.uint32).min
```

3. In seed_everything():
```
if _NUMPY_AVAILABLE:
    np.random.seed(seed)
```
4. In pl_worker_init_function():
```
if _NUMPY_AVAILABLE:
        ss = np.random.SeedSequence([base_seed, worker_id, global_rank])
        # use 128 bits (4 x 32-bit words)
        np.random.seed(ss.generate_state(4))
        # Spawn distinct SeedSequences for the PyTorch PRNG and the stdlib random module
        torch_ss, stdlib_ss = ss.spawn(2)
        torch.manual_seed(torch_ss.generate_state(1, dtype=np.uint64)[0])
        # use 128 bits expressed as an integer
        stdlib_seed = (stdlib_ss.generate_state(2, dtype=np.uint64).astype(object) * [1 << 64, 1]).sum()

 random.seed(stdlib_seed)

```

5. In _collect_rng_states():
```
def _collect_rng_states(include_cuda: bool = True) -> Dict[str, Any]:
    r"""Collect the global random state of :mod:`torch`, :mod:`torch.cuda`, :mod:`numpy` and Python."""
    states = {
        "torch": torch.get_rng_state(),
        "python": python_get_rng_state(),
    }
    if _NUMPY_AVAILABLE:
        states["numpy"] = np.random.get_state()
    if include_cuda:
        states["torch.cuda"] = torch.cuda.get_rng_state_all() if torch.cuda.is_available() else []
    return states
```
6. In _set_rng_states():
```
if _NUMPY_AVAILABLE and "numpy" in rng_state_dict:
    np.random.set_state(rng_state_dict["numpy"])
```

Below is the test I ran to ensure that numpy is an optional dependency now and all the functions are doing what they are supposed to do without numpy.

```
import unittest
import subprocess
import sys
import importlib
from unittest.mock import patch

# Command to uninstall numpy for the test
uninstall_numpy_command = [sys.executable, "-m", "pip", "uninstall", "-y", "numpy"]

# Command to reinstall numpy after the test
install_numpy_command = [sys.executable, "-m", "pip", "install", "numpy"]

class TestOptionalNumPyDependency(unittest.TestCase):
    
    @classmethod
    def setUpClass(cls):
        # Uninstall numpy if it is installed
        subprocess.run(uninstall_numpy_command, check=True)
    
    @classmethod
    def tearDownClass(cls):
        # Reinstall numpy after tests
        subprocess.run(install_numpy_command, check=True)
    
    def test_seed_everything_without_numpy(self):
        from lightning.fabric.utilities.seed import seed_everything

        # Mock logging to avoid unnecessary output during tests
        with patch('lightning.fabric.utilities.seed.log') as mock_log:
            # Test seed_everything without numpy
            seed = seed_everything(42, workers=True)
            self.assertEqual(seed, 42)
    
    def test_reset_seed_without_numpy(self):
        from lightning.fabric.utilities.seed import seed_everything, reset_seed

        # Mock logging to avoid unnecessary output during tests
        with patch('lightning.fabric.utilities.seed.log') as mock_log:
            # Test reset_seed without numpy
            seed_everything(42, workers=True)
            reset_seed()

    def test_pl_worker_init_function_without_numpy(self):
        from lightning.fabric.utilities.seed import pl_worker_init_function

        # Mock logging to avoid unnecessary output during tests
        with patch('lightning.fabric.utilities.seed.log') as mock_log:
            # Test pl_worker_init_function without numpy
            pl_worker_init_function(0)

if __name__ == '__main__':
    unittest.main()
```
tagging @awaelchli for better visibility.

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--20055.org.readthedocs.build/en/20055/

### CODE REVIEW

**Comment by 01AbhiSingh** [comment]
I accidently push the test file also with the PR. I have deleted the file though.

File: src/lightning/fabric/utilities/seed.py [hunk 1]
```diff
@@ -7,13 +7,18 @@
 
 import numpy as np
 import torch
+from lightning_utilities.core.imports import RequirementCache
 
 from lightning.fabric.utilities.rank_zero import _get_rank, rank_prefixed_message, rank_zero_only, rank_zero_warn
 
 log = logging.getLogger(__name__)
+_NUMPY_AVAILABLE = RequirementCache("numpy")
```
Comment (awaelchli) [inline src/lightning/fabric/utilities/seed.py:15]: We keep an inventory of these variables in fabric/utilities/imports.py, let's move it there and import it here :)

File: src/lightning/fabric/utilities/seed.py [hunk 2]
```diff
@@ -7,13 +7,18 @@
...
 
-max_seed_value = np.iinfo(np.uint32).max
-min_seed_value = np.iinfo(np.uint32).min
+if _NUMPY_AVAILABLE:
+    import numpy as np
+
+    max_seed_value = np.iinfo(np.uint32).max
+    min_seed_value = np.iinfo(np.uint32).min
```
Comment (awaelchli) [inline src/lightning/fabric/utilities/seed.py:21]: I suggest to actually just write these values as Python integers. We can write in a comment how we computed them:

```py
max_seed_value = 4294967295  # 2^32 - 1 (uint32)
min_seed_value = 0
```

File: src/lightning/fabric/utilities/seed.py [hunk 3]
```diff
@@ -91,24 +97,26 @@ def pl_worker_init_function(worker_id: int, rank: Optional[int] = None) -> None:
...
+    if _NUMPY_AVAILABLE:
+        ss = np.random.SeedSequence([base_seed, worker_id, global_rank])
+        # use 128 bits (4 x 32-bit words)
+        np.random.seed(ss.generate_state(4))
+        # Spawn distinct SeedSequences for the PyTorch PRNG and the stdlib random module
+        torch_ss, stdlib_ss = ss.spawn(2)
+        torch.manual_seed(torch_ss.generate_state(1, dtype=np.uint64)[0])
+        # use 128 bits expressed as an integer
+        stdlib_seed = (stdlib_ss.generate_state(2, dtype=np.uint64).astype(object) * [1 << 64, 1]).sum()
     random.seed(stdlib_seed)
```
Comment (awaelchli) [inline src/lightning/fabric/utilities/seed.py:109]: This here wouldn't be defined if numpy wasn't available. Also we still need to call torch.manual_seed() with a seed even if numpy is not available. This is a tricky part that we need to rewrite. Let me think about this for a bit to find a solution

**Review by awaelchli — COMMENTED** [review]
Yeah this is great, exactly what we need.

File: src/lightning/fabric/utilities/seed.py (hunk 3 above)
Comment (01AbhiSingh) [reply src/lightning/fabric/utilities/seed.py]:
```
if _NUMPY_AVAILABLE:
        ss = np.random.SeedSequence([base_seed, worker_id, global_rank])
        np.random.seed(ss.generate_state(4))
        torch_ss, stdlib_ss = ss.spawn(2)
        torch.manual_seed(torch_ss.generate_state(1, dtype=np.uint64)[0])
        stdlib_seed = (stdlib_ss.generate_state(2, dtype=np.uint64).astype(object) * [1 << 64, 1]).sum()
        random.seed(stdlib_seed)
    else:
        torch.manual_seed(base_seed)
        random.seed(base_seed)
```
@awaelchli  can you Please check this one good enough ??

File: src/lightning/fabric/utilities/seed.py (hunk 1 above)
Comment (01AbhiSingh) [reply src/lightning/fabric/utilities/seed.py]: Suggestion mentioned above implemented, Please check.

**Comment by 01AbhiSingh** [comment]
Hii @awaelchli.
 I was working on  remaining issue :
[Parameter sanitization](https://github.com/Lightning-AI/pytorch-lightning/blob/90f23b34e956ac19eda4f5b41b9d1339595fd207/src/lightning/fabric/utilities/logger.py#L133-L134)

you mentioned in this comment:

https://github.com/Lightning-AI/pytorch-lightning/issues/16649#issuecomment-2209526428

Wanted to ask that if  I am understanding the code correctly, the code below already check whether the value associated is a numpy scalar type or not 
```
CONVERSION_DTYPES: List[Tuple[Any, Callable[[Any, Any], Tensor]]] = [
    # bool -> uint8 as bool -> torch.bool triggers RuntimeError: Unsupported data type for NCCL process group
    (bool, partial(torch.tensor, dtype=torch.uint8)),
    (int, partial(torch.tensor, dtype=torch.int)),
    (float, partial(torch.tensor, dtype=torch.float)),
    (np.ndarray, _from_numpy),
]
```
If there is already a condition is there that checks the numpy is there or not, do we need to add another conditional statement there ?

Please help.  :)

File: src/lightning/fabric/utilities/seed.py (hunk 3 above)
Comment (awaelchli) [reply src/lightning/fabric/utilities/seed.py]: It's very close, but we will need an implementation for generating the random state like we did before, but without numpy (with torch instead). I can provide that implementation, or you can try it yourself too of course. I'll reserve some time to work on this and get back to you.

**Comment by awaelchli** [comment]
> My question is that if we already have a Numpy checking condition, is there a need to add another conditional statement to check whether numpy is available or not ?


@01AbhiSingh Yes there is still some work needed there. But it's much easier than the seed work here.
All you need to do is remove the import at the top of the file and move it locally to that function:


```py
if _NUMPY_AVAILABLE:
	import numpy as np 
	
	if isinstance(params[k], (np.bool_, np.integer, np.floating)):
		params[k] = params[k].item()
```

For this one, please send a separate PR, and we can discuss any additional challenges there :)
Thanks

**Comment by 01AbhiSingh** [comment]
```
if _NUMPY_AVAILABLE:
	import numpy as np 
	
	if isinstance(params[k], (np.bool_, np.integer, np.floating)):
		params[k] = params[k].item()
```
I will push this in a new PR.

What about this current PR ? Do I need to change anything in this current one or is it's all good ?

**Comment by 01AbhiSingh** [comment]
> Great work @01AbhiSingh! I also finally found the time to put together the alternative implementation for the seed sequence generator that was needed to replace the numpy implementation. I committed that to this branch directly and added a simple test.

It took me a little time to understand the difference between the code changes you made and my code change in the seed generation method lol . Definitely some high quality code I can learn a lot from.

**Pull request merged**